ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+  New += sizeof (SmmuConfig->Config.RcNode);
+
+  Status = AcpiPlatformChecksum ((UINT8 *)(UINTN)PageAddress, TableSize);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to calculate checksum for IORT table\n"));
+    return Status;
+  }
+
+  Status = AcpiTable->InstallAcpiTable (
+                        AcpiTable,
+                        (EFI_ACPI_COMMON_HEADER *)(UINTN)PageAddress,
+                        TableSize,
+                        &TableHandle
+                        );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to install IORT table\n"));
+  }
+
+  return Status;
+}
+
+/**
+  Allocate an event queue for SMMUv3.
+
+  @param [in]  SmmuInfo       Pointer to the SMMU_INFO structure.
+  @param [out] QueueLog2Size  Pointer to store the log2 size of the queue.
+
+  @retval Pointer to the allocated event queue.
+**/
+VOID *
+EFIAPI
+SmmuV3AllocateEventQueue (
+  IN SMMU_INFO  *SmmuInfo,
+  OUT UINT32    *QueueLog2Size
+  )
+{
+  UINT32       QueueSize;
+  SMMUV3_IDR1  Idr1;
+
+  Idr1.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_IDR1);
+
+  *QueueLog2Size = MIN (Idr1.EventQs, SMMUV3_EVENT_QUEUE_LOG2ENTRIES);
+  QueueSize      = SMMUV3_EVENT_QUEUE_SIZE_FROM_LOG2 (*QueueLog2Size);
+  return AllocateZeroPool (QueueSize);
+}
+
+/**
+  Allocate a command queue for SMMUv3.
+
+  @param [in]  SmmuInfo       Pointer to the SMMU_INFO structure.
+  @param [out] QueueLog2Size  Pointer to store the log2 size of the queue.
+
+  @retval Pointer to the allocated command queue.
+**/
+VOID *
+EFIAPI
+SmmuV3AllocateCommandQueue (
+  IN SMMU_INFO  *SmmuInfo,
+  OUT UINT32    *QueueLog2Size
+  )
+{
+  UINT32       QueueSize;
+  SMMUV3_IDR1  Idr1;
+
+  Idr1.AsUINT32 = SmmuV3ReadRegister32 (SmmuInfo->SmmuBase, SMMU_IDR1);
+
+  *QueueLog2Size = MIN (Idr1.CmdQs, SMMUV3_COMMAND_QUEUE_LOG2ENTRIES);
+  QueueSize      = SMMUV3_COMMAND_QUEUE_SIZE_FROM_LOG2 (*QueueLog2Size);
+  return AllocateZeroPool (QueueSize);
+}
+
+/**
+  Free a previously allocated queue.
+
+  @param [in]  QueuePtr    Pointer to the queue to free.
+**/
+VOID
+EFIAPI
+SmmuV3FreeQueue (
+  IN VOID  *QueuePtr
+  )
+{
+  FreePool (QueuePtr);
+}
+
+/**
+  Build the stream table for SMMUv3.
+
+  @param [in]  SmmuInfo       Pointer to the SMMU_INFO structure.
+  @param [in]  SmmuConfig     Pointer to the SMMU configuration.
+  @param [out] StreamEntry    Pointer to the stream table entry.
+
+  @retval EFI_SUCCESS         Success.
+  @retval EFI_INVALID_PARAMETER  Invalid parameter.
+**/
+EFI_STATUS
+EFIAPI
+SmmuV3BuildStreamTable (
+  IN SMMU_INFO                   *SmmuInfo,
+  IN SMMU_CONFIG                 *SmmuConfig,
+  OUT SMMUV3_STREAM_TABLE_ENTRY  *StreamEntry
+  )
+{
+  EFI_STATUS   Status;
+  UINT32       OutputAddressWidth;
+  UINT32       InputSize;
+  SMMUV3_IDR0  Idr0;
+  SMMUV3_IDR1  Idr1;
+  SMMUV3_IDR5  Idr5;
+
+  UINT8   IortCohac = SmmuConfig->Config.SmmuNode.SmmuNode.Flags & EFI_ACPI_IORT_SMMUv3_FLAG_COHAC_OVERRIDE;

```

Comment:
Check `SmmuConfig` before using it?

Also, the edk2 style is to define all variables and assign later.

Reply:
both done.